### PR TITLE
EVG-7645: report handled and unhandled errors to bugsnag

### PR DESCRIPTION
**Pull request by SupaJoon**
https://jira.mongodb.org/browse/EVG-7645

### CODE REVIEW

File: src/pages/commitqueue/CommitQueueCard.tsx [hunk 1]
```diff
@@ -41,16 +41,12 @@ export const CommitQueueCard: React.FC<Props> = ({
...
-        variables: { patchId, commitQueueId },
-        refetchQueries: ["CommitQueue"],
-      });
-    } catch (err) {
-      // console.log(err); // TODO: Replace this with better error handling
-    }
+    removePatchFromCommitQueue({
+      variables: { patchId, commitQueueId },
+      refetchQueries: ["CommitQueue"],
+    });
```
Comment (tgrander) [inline src/pages/commitqueue/CommitQueueCard.tsx:49]: if a mutation errors and the promise is not caught it crashes the app; i just tested it by trying to configure a patch locally, which always errors. the app crashed. all mutations need to be wrapped in a try/catch

File: src/pages/configurePatch/ConfigurePatchCore.tsx [hunk 2]
```diff
@@ -60,19 +60,14 @@ export const ConfigurePatchCore: React.FC<Props> = ({ patch }) => {
...
-      await schedulePatch({
-        variables: { patchId: id, reconfigure: configurePatchParam },
-      });
-    } catch (error) {
-      // TODO: log this error
-      // console.error(error);
-    }
+    schedulePatch({
+      variables: { patchId: id, reconfigure: configurePatchParam },
+    });
```
Comment (tgrander) [inline src/pages/configurePatch/ConfigurePatchCore.tsx:70]: same comment about being wrapped in a try/catch

**Comment by tgrander** [comment]
Is there going to be a follow up PR to call bugsnag for all errors in the application?

**Comment by SupaJoon** [comment]
> Is there going to be a follow up PR to call bugsnag for all errors in the application?

the Bugsnag error boundary reports all unhandled errors and the errorsLink in the GQLWrapper reports all GQLErrors. I called reportError on all errors we throw and catch with our non bugsnag ErrorBoundary but I decided not to call reportError for errors thrown when misusing context because those are obvious as we are developing. Non graphql network errors are handled individually where they are called. I think this covers all of them.

**Comment by SupaJoon** [comment]
> i am weary about this magically capturing all the errors we need, but i trust what you tell me and so am approving. can we confirm that all bugs are indeed being captured when this is in production?

I confirmed that the bugsnag error boundary works in staging with the bugsnag api key. i tested it by throwing an unhandled error in our app and making sure that it was logged to bugsnag. 
Screen Shot 2020-06-05 at 2 16 57 PM

File: src/components/TaskStatusBadge.tsx [hunk 3]
```diff
@@ -60,5 +61,7 @@ export const TaskStatusBadge: React.FC<{ status: string }> = ({ status }) => {
       </StyledBadge>
     );
   }
-  throw new Error(`Status '${status}' is not a valid task status`);
+  const err = new Error(`Status '${status}' is not a valid task status`);
+  reportError(err).severe();
```
[the remark this answers is not in the document]
Comment (SupaJoon) [reply src/components/TaskStatusBadge.tsx:65]: this means we are receiving data we are unsure how to handle and I want it to stand out from network request errors

File: src/gql/GQLWrapper.tsx [hunk 4]
```diff
@@ -72,6 +73,16 @@ const authLink = (logout: Logout): ApolloLink =>
...
   });
 
+const logErrorsLink = onError(({ graphQLErrors }) => {
+  if (Array.isArray(graphQLErrors)) {
+    graphQLErrors.forEach((gqlErr) => {
+      reportError(gqlErr).warning();
+    });
+  }
+  // dont track network errors here because they are
+  // very common when a user is not authenticated
```
[the remark this answers is not in the document]
Comment (SupaJoon) [reply src/gql/GQLWrapper.tsx:83]: I log errors for all of our network requests that we use in production for spruce. I don't think we will request anything we didn't mean to without us noticing.

File: src/pages/patch/patchTabs/Tasks.tsx [hunk 5]
```diff
@@ -175,7 +175,11 @@ const renderStatusBadge = (status): null | JSX.Element => {
   if (status === "" || !status) {
     return null;
   }
-  return <TaskStatusBadge status={status} />;
+  return (
+    <ErrorBoundary>
```
[the remark this answers is not in the document]
Comment (SupaJoon) [reply src/pages/patch/patchTabs/Tasks.tsx:179]: good point. just made the change

**Pull request merged**